File: novel-app/server/src/routes/novel.ts
// 小说核心路由：目录、创建、大纲、章节、状态。
// 对应 commands.rs 的 novel 相关命令 + start_outline_generation 后台任务状态管理。

import type { FastifyInstance } from "fastify";

import type { Prompts } from "../config.js";
import { AppError } from "../error.js";
import { listChapters, readChapter } from "../files/index.js";
import { configFromState, dirFromState } from "../lib/state-helpers.js";
import { emitOutlineGenerationStatus } from "../lib/sse.js";
import {
  createNovel,
  generateOutlineStreamingWithProgress,
} from "../logic/outline.js";
import { generateChapterStreaming } from "../logic/chapter.js";
import { getStatus } from "../logic/status.js";
import { getState } from "../state.js";

export function registerNovelRoutes(app: FastifyInstance): void {
  // select_novel_dir：前端通过 Tauri 对话框拿到路径后 POST 给 Node
  // body: { path: string }
  app.post("/api/select_novel_dir", async (req) => {
    const { path: dirPath } = (req.body ?? {}) as { path?: string };
    if (!dirPath || dirPath === "") {
      throw AppError.noNovelDir();
    }
    getState().novelDir = dirPath;
    // 持久化到配置（照搬 commands.rs:64-67）
    const { loadConfig, saveConfig } = await import("../config.js");
    const config = loadConfig(getState().configPath);
    config.novel_dir = dirPath;
    saveConfig(getState().configPath, config);
    return dirPath;
  });

  // get_novel_dir
  app.get("/api/get_novel_dir", () => getState().novelDir);

  // create_novel
  // body: { title, genre, theme, chapters, overwrite?, promptsOverride? }
  app.post("/api/create_novel", async (req) => {
    const dir = dirFromState();
    const config = configFromState();
    const body = (req.body ?? {}) as {
      title: string;
      genre: string;
      theme: string;
      chapters: number;
      overwrite?: boolean;
      promptsOverride?: Prompts;
    };
    const effectiveConfig =
      body.promptsOverride !== undefined
        ? { ...config, prompts: body.promptsOverride }
        : config;
    createNovel(
      dir,
      body.title,
      body.genre,
      body.theme,
      body.chapters,
      effectiveConfig,
      body.overwrite ?? false,
    );
    return null;
  });

  // generate_outline（同步全流程，返回 outline_text）
  app.post("/api/generate_outline", async () => {
    const dir = dirFromState();
    const config = configFromState();
    return generateOutlineStreamingWithProgress(dir, config, "", () => {});
  });

  // start_outline_generation：启动后台任务，立即返回。状态通过 get_outline_generation_status 查询。
  // body: { step?: string }
  app.post("/api/start_outline_generation", async (req) => {
    const dir = dirFromState();
    const config = configFromState();
    const body = (req.body ?? {}) as { step?: string };
    const targetStep = body.step ?? "";

    const state = getState();
    // 幂等：已在运行则直接返回（照搬 commands.rs:124-128）
    if (state.outlineGeneration.running) {
      return null;
    }
    state.outlineGeneration = {
      running: true,
      completed: false,
      currentStep: targetStep === "" ? undefined : (targetStep as "world" | "characters" | "outline"),
      streamingText: {},
      error: undefined,
    };

    // 后台执行（不 await）
    void runOutlineGeneration(dir, config, targetStep);

    return null;
  });

  // get_outline_generation_status
  app.get("/api/get_outline_generation_status", () => getState().outlineGeneration);

  // generate_chapter
  app.post("/api/generate_chapter", async () => {
    const dir = dirFromState();
    const config = configFromState();
    return generateChapterStreaming(dir, config);
  });

  // get_status
  app.get("/api/get_status", () => getStatus(dirFromState()));

  // list_chapters
  app.get("/api/list_chapters", () => listChapters(dirFromState()));

  // read_chapter?filename=xxx
  app.get("/api/read_chapter", async (req) => {
    const { filename } = req.query as { filename: string };
    const dir = dirFromState();
    return readChapter(dir, filename);
  });
}

// 后台大纲生成任务，照搬 commands.rs:139-170 的 spawn 逻辑。
// 完成时更新状态并推送 outline-generation-status 事件。
async function runOutlineGeneration(
  dir: string,
  config: ReturnType<typeof configFromState>,
  targetStep: string,
): Promise<void> {
  const state = getState();
  try {
    await generateOutlineStreamingWithProgress(
      dir,
      config,
      targetStep,
      (step, chunk, _done) => {
        // 更新流式文本状态（照搬 commands.rs:147-156）
        const s = state.outlineGeneration;
        s.currentStep = step as "world" | "characters" | "outline";
        if (chunk !== "") {
          s.streamingText[step as "world" | "characters" | "outline"] =
            (s.streamingText[step as "world" | "characters" | "outline"] ?? "") + chunk;
        }
      },
    );
    // 成功
    state.outlineGeneration.running = false;
    state.outlineGeneration.completed = true;
    state.outlineGeneration.error = undefined;
  } catch (e) {
    state.outlineGeneration.running = false;
    state.outlineGeneration.completed = false;
    const { errorToString, normalizeError } = await import("../error.js");
    state.outlineGeneration.error = errorToString(normalizeError(e));
  }
  // 推送最终状态快照（照搬 commands.rs:169）
  emitOutlineGenerationStatus({ ...state.outlineGeneration });
}

File: app/app-tauri/src/commands/novel.rs
use std::collections::HashMap;

use autowrite_core::domain::util::map_step;
use autowrite_core::progress::{OutlineStep, ProgressEvent};
use tauri::{Emitter, Manager, State};

use super::{config_from_state, dir_from_state};
use crate::domain::novel;
use crate::dto::{ChapterContentDto, ChapterMetaDto, NovelStatusDto, OutlineGenerationStatusDto};
use crate::error::Result;
use crate::state::{AppState, OutlineGenerationStatus};

/// IPC 事件载荷：大纲生成流式进度（ADR-009 §3.1.4，留在 app 侧）。
#[derive(Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
struct OutlineProgressEvent {
    step: String,
    chunk: String,
    done: bool,
}

/// IPC 事件载荷：章节生成流式进度（ADR-009 §3.1.4，留在 app 侧）。
#[derive(Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
struct ChapterProgressEvent {
    chunk: String,
    done: bool,
}

/// 把 core 的 OutlineStep 映射回后端内部步骤名（与提示词模板、AppState key 对齐）。
/// AppState 的 streaming_text 用内部名（world/characters/outline）做 key，
/// emit 事件再经 map_step 映射成前端契约值（worldView/characters/outline）。
fn outline_step_internal(step: OutlineStep) -> &'static str {
    match step {
        OutlineStep::World => "world",
        OutlineStep::Characters => "characters",
        OutlineStep::Outline => "outline",
    }
}

#[tauri::command]
pub async fn create_novel(
    state: State<'_, AppState>,
    title: String,
    genre: String,
    theme: String,
    chapters: u32,
    overwrite: Option<bool>,
    prompts_override: Option<crate::domain::config::Prompts>,
) -> Result<()> {
    let dir = dir_from_state(&state)?;
    let mut config = config_from_state(&state)?;
    if let Some(p) = prompts_override {
        config.prompts = p;
    }
    novel::create_novel(
        &dir,
        &title,
        &genre,
        &theme,
        chapters,
        &config,
        overwrite.unwrap_or(false),
    )
}

#[tauri::command]
pub async fn generate_outline(state: State<'_, AppState>) -> Result<String> {
    let dir = dir_from_state(&state)?;
    let config = config_from_state(&state)?;
    novel::generate_outline_streaming(&dir, &config).await
}

#[tauri::command]
pub async fn start_outline_generation(
    app: tauri::AppHandle,
    state: State<'_, AppState>,
    step: Option<String>,
) -> Result<()> {
    let dir = dir_from_state(&state)?;
    let config = config_from_state(&state)?;
    let target_step = step.unwrap_or_default();

    {
        let mut status = state
            .outline_generation
            .lock()
            .unwrap_or_else(|e| e.into_inner());
        if status.running {
            return Ok(());
        }
        *status = OutlineGenerationStatus {
            running: true,
            completed: false,
            current_step: Some(target_step.clone()),
            streaming_text: HashMap::new(),
            error: None,
        };
    }

    let app_for_task = app.clone();
    tauri::async_runtime::spawn(async move {
        let app_for_progress = app_for_task.clone();
        // 注入闭包（ADR-009 §3.1.4）：core 产出中性的 ProgressEvent，
        // 闭包内更新 AppState 并还原原 emit 语义。
        let result = novel::generate_outline_streaming_with_progress(
            &dir,
            &config,
            &target_step,
            move |ev: ProgressEvent| {
                if let ProgressEvent::OutlineStep { step, chunk, done } = ev {
                    let internal = outline_step_internal(step);
                    let contract = map_step(internal);

                    // 1. 更新 AppState（用内部步骤名做 streaming_text 的 key）
                    let state = app_for_progress.state::<AppState>();
                    let mut status = state
                        .outline_generation
                        .lock()
                        .unwrap_or_else(|e| e.into_inner());
                    status.current_step = Some(internal.to_string());
                    if !chunk.is_empty() {
                        status
                            .streaming_text
                            .entry(internal.to_string())
                            .or_default()
                            .push_str(&chunk);
                    }

                    // 2. 还原原 emit（OutlineProgressEvent 用前端契约步骤名）
                    let _ = app_for_progress.emit(
                        "outline-progress",
                        OutlineProgressEvent {
                            step: contract.to_string(),
                            chunk,
                            done,
                        },
                    );
                }
            },
        )
        .await;

        let state = app_for_task.state::<AppState>();
        let status_snapshot = {
            let mut status = state
                .outline_generation
                .lock()
                .unwrap_or_else(|e| e.into_inner());
            status.running = false;
            status.completed = result.is_ok();
            status.error = result.err().map(|e| e.to_string());
            status.clone()
        };
        // 经 DTO 转换：步骤名 "world" → "worldView"，与轮询契约一致。
        let snapshot_dto: OutlineGenerationStatusDto = status_snapshot.into();
        let _ = app_for_task.emit("outline-generation-status", snapshot_dto);
    });

    Ok(())
}

#[tauri::command]
pub fn get_outline_generation_status(
    state: State<'_, AppState>,
) -> Result<OutlineGenerationStatusDto> {
    Ok(state
        .outline_generation
        .lock()
        .unwrap_or_else(|e| e.into_inner())
        .clone()
        .into())
}

#[tauri::command]
pub async fn generate_chapter(app: tauri::AppHandle, state: State<'_, AppState>) -> Result<u32> {
    let dir = dir_from_state(&state)?;
    let config = config_from_state(&state)?;
    let app_for_emit = app.clone();
    crate::domain::chapter::generate_chapter_streaming(&dir, &config, move |ev: ProgressEvent| {
        if let ProgressEvent::ChapterChunk { chunk, done } = ev {
            let _ = app_for_emit.emit(
                "chapter-progress",
                ChapterProgressEvent { chunk, done },
            );
        }
    })
    .await
}

#[tauri::command]
pub fn get_status(state: State<'_, AppState>) -> Result<NovelStatusDto> {
    let dir = dir_from_state(&state)?;
    Ok(novel::get_status(&dir)?.into())
}

#[tauri::command]
pub fn list_chapters(state: State<'_, AppState>) -> Result<Vec<ChapterMetaDto>> {
    let dir = dir_from_state(&state)?;
    let ch_dir = crate::services::files::chapters_dir(&dir);
    if !ch_dir.exists() {
        return Ok(Vec::new());
    }

    // 领域 ChapterMeta 无 filename 字段，遍历目录取真实文件名注入 DTO。
    // 与 services::files::list_chapters 一致的解析逻辑，但保留文件名。
    let mut entries: Vec<ChapterMetaDto> = Vec::new();
    for entry in std::fs::read_dir(&ch_dir)? {
        let entry = entry?;
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("md") {
            continue;
        }
        let content = match std::fs::read_to_string(&path) {
            Ok(c) => c,
            Err(_) => continue,
        };
        let content = content.replace("\r\n", "\n");
        let (meta_yaml, _) = crate::services::files::parse_yaml_front_matter(&content);
        if meta_yaml.is_null() {
            continue;
        }
        let Ok(meta) = serde_yaml::from_value::<crate::domain::types::ChapterMeta>(meta_yaml)
        else {
            continue;
        };
        let filename = entry
            .file_name()
            .to_string_lossy()
            .to_string();
        entries.push(ChapterMetaDto::from_meta(meta, filename));
    }
    entries.sort_by_key(|c| c.number);
    Ok(entries)
}

#[tauri::command]
pub fn read_chapter(state: State<'_, AppState>, filename: String) -> Result<ChapterContentDto> {
    let dir = dir_from_state(&state)?;
    // 路径穿越校验（P0-3）：确保 filename 不含 .. 或绝对路径，且 join 后在 dir 内
    let chapters_base = crate::services::files::chapters_dir(&dir);
    super::safe_join(&chapters_base, &filename)?;
    let (meta, body) = crate::services::files::read_chapter(&dir, &filename)?;
    Ok(ChapterContentDto::from_parts(meta, filename, body))
}
